File: app/src/main/java/com/rtukpe/fixtures/ui/competition/teams/TeamsFragment.java
package com.rtukpe.fixtures.ui.competition.teams;


import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.View;
import android.view.ViewGroup;

import com.rtukpe.fixtures.R;
import com.rtukpe.fixtures.data.model.Team;
import com.rtukpe.fixtures.di.component.ActivityComponent;
import com.rtukpe.fixtures.ui.base.BaseFragment;
import com.rtukpe.fixtures.ui.competitions.CompetitionsFragment;
import com.rtukpe.fixtures.utils.others.ItemOffsetDecoration;

import java.util.ArrayList;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;

public class TeamsFragment extends BaseFragment implements TeamsMvpView {

    @BindView(R.id.recylcer_view)
    RecyclerView recyclerView;

    @Inject
    LinearLayoutManager mLinearLayoutManager;

    @Inject
    TeamsAdapter teamsAdapter;

    @Inject
    GridLayoutManager gridLayoutManager;

    @Inject
    TeamsMvpContract<TeamsMvpView> mPresenter;

    public TeamsFragment() {

    }

    public static TeamsFragment newInstance() {
        TeamsFragment fragment = new TeamsFragment();
        Bundle args = new Bundle();

        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_teams, container, false);
        ActivityComponent component = getActivityComponent();
        if (component != null) {
            component.inject(this);
            setUnBinder(ButterKnife.bind(this, view));
            mPresenter.onAttach(this);
            setHasOptionsMenu(true);
        }
        return view;
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        menu.clear();
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    protected void setUp(View view) {
        mPresenter.onViewInitialized();

        int id = getBaseActivity().getIntent().getIntExtra(CompetitionsFragment.COMPETITION_ID, 0);
        mPresenter.getTeams(id);

//        ListPreloader.PreloadSizeProvider<Team> sizeProvider = new FixedPreloadSizeProvider<>(64, 64);
//        BitmapPreloadModelProvider bitmapPreloadModelProvider = new BitmapPreloadModelProvider(getContext(), teamsAdapter.getTeams());
//        PictureDrawablePreloadModelProvider pictureDrawablePreloadModelProvider = new PictureDrawablePreloadModelProvider(getBaseActivity(), teamsAdapter.getTeams());
//        RecyclerViewPreloader<Team> preloader = new RecyclerViewPreloader<>(Glide.with(getBaseActivity()), pictureDrawablePreloadModelProvider, sizeProvider, 20 /*maxPreload*/);
//        recyclerView.addOnScrollListener(preloader);
        recyclerView.setAdapter(teamsAdapter);
        recyclerView.setLayoutManager(gridLayoutManager);
        ItemOffsetDecoration itemDecoration = new ItemOffsetDecoration(getBaseActivity(), R.dimen.item_offset);
        recyclerView.addItemDecoration(itemDecoration);
        teamsAdapter.setRecyclerViewClickListener(this);
    }

    @Override
    public void recyclerViewListClicked(View v, int position) {

    }

    @Override
    public void updateTeams(ArrayList<Team> teams) {
        teamsAdapter.addTeams(teams);
    }
}

File: app/src/main/java/com/rtukpe/fixtures/ui/competition/teams/TeamsFragment.kt
package com.rtukpe.fixtures.ui.competition.teams


import android.os.Bundle
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.*
import butterknife.BindView
import butterknife.ButterKnife
import com.rtukpe.fixtures.R
import com.rtukpe.fixtures.data.model.Team
import com.rtukpe.fixtures.ui.base.BaseFragment
import com.rtukpe.fixtures.ui.competition.teams.team.TeamBottomSheetFragment
import com.rtukpe.fixtures.ui.competitions.CompetitionsFragment
import com.rtukpe.fixtures.utils.others.ItemOffsetDecoration
import java.util.*
import javax.inject.Inject

class TeamsFragment : BaseFragment(), TeamsMvpView, TeamsAdapter.OnRetryClicked {

    @BindView(R.id.recylcer_view)
    lateinit var recyclerView: RecyclerView

    @Inject
    lateinit var mLinearLayoutManager: LinearLayoutManager

    @Inject
    lateinit var teamsAdapter: TeamsAdapter

    @Inject
    lateinit var gridLayoutManager: GridLayoutManager

    @Inject
    lateinit var mPresenter: TeamsMvpContract<TeamsMvpView>

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_teams, container, false)
        val component = activityComponent
        if (component != null) {
            component.inject(this)
            setUnBinder(ButterKnife.bind(this, view))
            mPresenter.onAttach(this)
            setHasOptionsMenu(true)
        }
        return view
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        menu?.clear()
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun setUp(view: View) {
        mPresenter.onViewInitialized()

        val id = baseActivity.intent.getIntExtra(CompetitionsFragment.COMPETITION_ID, 0)
        mPresenter.getTeams(id)

        recyclerView.adapter = teamsAdapter
        recyclerView.layoutManager = gridLayoutManager
        val itemDecoration = ItemOffsetDecoration(baseActivity, R.dimen.item_offset)
        recyclerView.addItemDecoration(itemDecoration)
        teamsAdapter.setRecyclerViewClickListener(this)
        teamsAdapter.onRetryClickedListener = this
    }

    override fun recyclerViewListClicked(v: View, position: Int) {
        val dialog = TeamBottomSheetFragment.newInstance(teamsAdapter.getTeamAtPosition(position))
        dialog.show(baseActivity.supportFragmentManager, "TEAM")
    }

    override fun updateTeams(fixtures: ArrayList<Team>) {
        teamsAdapter.addTeams(fixtures)
    }

    override fun onRetryClicked() {
        val id = baseActivity.intent.getIntExtra(CompetitionsFragment.COMPETITION_ID, 0)
        mPresenter.getTeams(id)
    }

    companion object {

        fun newInstance(): TeamsFragment {
            val fragment = TeamsFragment()
            val args = Bundle()

            fragment.arguments = args
            return fragment
        }
    }
}
